File: csharp/test/Vista.SDK.Tests/GmodTests.cs
using System.Diagnostics;
using FluentAssertions;
using Vista.SDK;
using Vista.SDK.Internal;

namespace Vista.SDK.Tests;

public class GmodTests
{
    public sealed record ExpectedValues(string Max, int Count);

    public static IEnumerable<object[]> Test_Vis_Versions => VisVersions.All.Select(x => new object[] { x }).ToArray();

    public static readonly Dictionary<VisVersion, ExpectedValues> ExpectedMaxes =
        new()
        {
            { VisVersion.v3_4a, new("C1053.3114", 6420) },
            { VisVersion.v3_5a, new("C1053.3114", 6557) },
            { VisVersion.v3_6a, new("C1053.3114", 6557) },
            { VisVersion.v3_7a, new("H346.11113", 6672) },
            { VisVersion.v3_8a, new("H346.11113", 6335) }
        };

    [Theory]
    [MemberData(nameof(Test_Vis_Versions))]
    public void Test_Gmod_Loads(VisVersion visVersion)
    {
        var (_, vis) = VISTests.GetVis();

        var gmod = vis.GetGmod(visVersion);
        Assert.NotNull(gmod);

        Assert.True(gmod.TryGetNode("400a", out _));
    }

    [Theory]
    [MemberData(nameof(Test_Vis_Versions))]
    public void Test_Gmod_Properties(VisVersion visVersion)
    {
        // This test ensures certain properties of the Gmod data
        // that we make some design desicisions based on,
        // i.e for hashing of the node code

        var (_, vis) = VISTests.GetVis();

        var gmod = vis.GetGmod(visVersion);
        Assert.NotNull(gmod);

        GmodNode? minLengthNode = null;
        int currentMinLength = int.MaxValue;

        GmodNode? maxLengthNode = null;
        int currentMaxLength = 0;
        int nodeCount = 0;

        foreach (var node in gmod)
        {
            nodeCount++;
            string code = node.Code;
            int len = code.Length;

            if (minLengthNode is null || len < currentMinLength)
            {
                currentMinLength = len;
                minLengthNode = node;
            }

            if (maxLengthNode is null || len > currentMaxLength)
            {
                currentMaxLength = len;
                maxLengthNode = node;
            }
            else if (len == currentMaxLength)
            {
                if (maxLengthNode is not null && string.CompareOrdinal(code, maxLengthNode.Code) > 0)
                {
                    maxLengthNode = node;
                }
            }
        }

        Assert.NotNull(minLengthNode);
        Assert.NotNull(maxLengthNode);

        Assert.Equal(2, minLengthNode.Code.Length);
        Assert.Equal("VE", minLengthNode.Code);

        Assert.Equal(10, maxLengthNode.Code.Length);
        Assert.True(ExpectedMaxes.TryGetValue(visVersion, out var expectedValues));

        Assert.Equal(expectedValues.Max, maxLengthNode.Code);

        Assert.Equal(expectedValues.Count, nodeCount);
    }

    [Theory]
    [MemberData(nameof(Test_Vis_Versions))]
    public void Test_Gmod_Lookup(VisVersion visVersion)
    {
        var (_, vis) = VISTests.GetVis();

        var gmod = vis.GetGmod(visVersion);
        Assert.NotNull(gmod);

        var gmodDto = VIS.Instance.GetGmodDto(visVersion);
        Assert.NotNull(gmodDto);

        {
            var seen = new HashSet<string>(StringComparer.Ordinal);
            var counter = 0;
            foreach (var node in gmodDto.Items)
            {
                Assert.NotNull(node?.Code);
                Assert.True(seen.Add(node.Code), $"Code: {node.Code}");

                Assert.NotNull(node.Code);
                Assert.True(gmod.TryGetNode(node.Code.AsSpan(), out var foundNode));
                Assert.NotNull(foundNode);
                Assert.Equal(node.Code, foundNode.Code);
                counter++;
            }
        }

        {
            var seen = new HashSet<string>(StringComparer.Ordinal);
            var counter = 0;
            foreach (var node in gmod)
            {
                Assert.NotNull(node?.Code);
                Assert.True(seen.Add(node.Code), $"Code: {node.Code}");

                Assert.NotNull(node.Code);
                Assert.True(gmod.TryGetNode(node.Code.AsSpan(), out var foundNode));
                Assert.NotNull(foundNode);
                Assert.Equal(node.Code, foundNode.Code);
                counter++;
            }

            Assert.Equal(gmodDto.Items.Length, counter);
        }

        Assert.False(gmod.TryGetNode("ABC", out _));
        Assert.False(gmod.TryGetNode(null!, out _));
        Assert.False(gmod.TryGetNode("", out _));
        Assert.False(gmod.TryGetNode("SDFASDFSDAFb", out _));
        Assert.False(gmod.TryGetNode("✅", out _));
        Assert.False(gmod.TryGetNode("a✅b", out _));
        Assert.False(gmod.TryGetNode("ac✅bc", out _));
        Assert.False(gmod.TryGetNode("✅bc", out _));
        Assert.False(gmod.TryGetNode("a✅", out _));
        Assert.False(gmod.TryGetNode("ag✅", out _));
    }

    [Fact]
    public void Test_Gmod_Node_Equality()
    {
        var (_, vis) = VISTests.GetVis();

        var gmod = vis.GetGmod(VisVersion.v3_4a);

        var node1 = gmod["400a"];

        var node2 = gmod["400a"];

        Assert.Equal(node1, node2);
        Assert.Same(node1, node2);

        var node3 = node2.WithLocation("1");
        Assert.NotEqual(node1, node3);
        Assert.NotSame(node1, node3);

        var node4 = node2 with { };
        Assert.Equal(node1, node4);
        Assert.NotSame(node1, node4);
    }

    [Fact]
    public void Test_Gmod_Node_Types()
    {
        var (_, vis) = VISTests.GetVis();

        var gmod = vis.GetGmod(VisVersion.v3_4a);

        var set = new HashSet<string>();
        foreach (var node in gmod)
            set.Add($"{node.Metadata.Category} | {node.Metadata.Type}");

        Assert.NotEmpty(set);
    }

    [Theory]
    [MemberData(nameof(Test_Vis_Versions))]
    public void Test_Gmod_RootNode_Children(VisVersion visVersion)
    {
        var (_, vis) = VISTests.GetVis();

        var gmod = vis.GetGmod(visVersion);

        var node = gmod.RootNode;

        Assert.NotEmpty(node.Children);
    }

    [Fact]
    public void Test_Normal_Assignments()
    {
        var (_, vis) = VISTests.GetVis();

        var gmod = vis.GetGmod(VisVersion.v3_4a);

        var node = gmod["411.3"];
        Assert.NotNull(node.ProductType);
        Assert.Null(node.ProductSelection);

        node = gmod["H601"];
        Assert.Null(node.ProductType);
    }

    [Fact]
    public void Test_Node_With_Product_Selection()
    {
        var (_, vis) = VISTests.GetVis();

        var gmod = vis.GetGmod(VisVersion.v3_4a);

        var node = gmod["411.2"];
        Assert.NotNull(node.ProductSelection);
        Assert.Null(node.ProductType);

        node = gmod["H601"];
        Assert.Null(node.ProductSelection);
    }

    [Fact]
    public void Test_Product_Selection()
    {
        var (_, vis) = VISTests.GetVis();

        var gmod = vis.GetGmod(VisVersion.v3_4a);

        var node = gmod["CS1"];
        Assert.True(node.IsProductSelection);
    }

    [Theory]
    [InlineData("VE", false)]
    [InlineData("300a", false)]
    [InlineData("300", true)]
    [InlineData("411", true)]
    [InlineData("410", true)]
    [InlineData("651.21s", false)]
    [InlineData("924.2", true)]
    [InlineData("411.1", false)]
    [InlineData("C101", true)]
    [InlineData("CS1", false)]
    [InlineData("C101.663", true)]
    [InlineData("C101.4", true)]
    [InlineData("C101.21s", false)]
    [InlineData("F201.11", true)]
    [InlineData("C101.211", false)]
    public void Test_Mappability(string code, bool mappable)
    {
        var (_, vis) = VISTests.GetVis();

        var gmod = vis.GetGmod(VisVersion.v3_4a);

        var node = gmod[code];

        Assert.Equal(mappable, node.IsMappable);
    }

    [Fact]
    public void Test_Full_Traversal()
    {
        var (_, vis) = VISTests.GetVis();

        var gmod = vis.GetGmod(VisVersion.v3_4a);

        var paths = new List<GmodPath>();
        var maxExpected = TraversalOptions.DEFAULT_MAX_TRAVERSAL_OCCURRENCE;
        int maxOccurrence = 0;
        var completed = gmod.Traverse(
            (parents, node) =>
            {
                Assert.True(parents.Count == 0 || parents[0].IsRoot);

                if (parents.Any(p => p.Code == "HG3") || node.Code == "HG3")
                {
                    paths.Add(new GmodPath(parents.ToList(), node));
                }

                var skipOccurenceCheck = Gmod.IsProductSelectionAssignment(parents.LastOrDefault(), node);
                if (skipOccurenceCheck)
                    return TraversalHandlerResult.Continue;
                var occ = Occurrences(parents, node);
                if (occ > maxOccurrence)
                    maxOccurrence = occ;

                return TraversalHandlerResult.Continue;
            }
        );
        Assert.Equal(maxExpected, maxOccurrence);
        Assert.True(completed);
    }

    [Fact]
    public void Test_Full_Traversal_With_Options()
    {
        var (_, vis) = VISTests.GetVis();

        var gmod = vis.GetGmod(VisVersion.v3_4a);

        var maxExpected = 2;
        int maxOccurrence = 0;
        var completed = gmod.Traverse(
            (parents, node) =>
            {
                var skipOccurenceCheck = Gmod.IsProductSelectionAssignment(parents.LastOrDefault(), node);
                if (skipOccurenceCheck)
                    return TraversalHandlerResult.Continue;
                var occ = Occurrences(parents, node);
                if (occ > maxOccurrence)
                    maxOccurrence = occ;
                return TraversalHandlerResult.Continue;
            },
            new TraversalOptions { MaxTraversalOccurrence = maxExpected }
        );
        Assert.Equal(maxExpected, maxOccurrence);
        Assert.True(completed);
    }

    [Fact]
    public void Test_Partial_Traversal()
    {
        var (_, vis) = VISTests.GetVis();

        var gmod = vis.GetGmod(VisVersion.v3_4a);

        var state = new TraversalState(5) { NodeCount = 0 };

        var completed = gmod.Traverse(
            state,
            (state, parents, node) =>
            {
                Assert.True(parents.Count == 0 || parents[0].IsRoot);
                if (++state.NodeCount == state.StopAfter)
                    return TraversalHandlerResult.Stop;
                return TraversalHandlerResult.Continue;
            }
        );

        Assert.Equal(state.StopAfter, state.NodeCount);
        Assert.False(completed);
    }

    [Fact]
    public void Test_Full_Traversal_From()
    {
        var (_, vis) = VISTests.GetVis();

        var gmod = vis.GetGmod(VisVersion.v3_4a);

        var state = new TraversalState(0) { NodeCount = 0 };

        var completed = gmod.Traverse(
            state,
            gmod["400a"],
            (state, parents, node) =>
            {
                Assert.True(parents.Count == 0 || parents[0].Code == "400a");
                ++state.NodeCount;
                return TraversalHandlerResult.Continue;
            }
        );
        Assert.True(completed);
    }

    private sealed record TraversalState(int StopAfter)
    {
        public int NodeCount { get; set; }
    }

    private int Occurrences(IReadOnlyList<GmodNode> parents, GmodNode node) => parents.Count(p => p.Code == node.Code);
}

File: cpp/test/GmodTests.cpp
#include "pch.h"

#include "dnv/vista/sdk/VIS.h"
#include "dnv/vista/sdk/VisVersion.h"
#include "dnv/vista/sdk/Gmod.h"
#include "dnv/vista/sdk/GmodDto.h"
#include "dnv/vista/sdk/GmodNode.h"
#include "dnv/vista/sdk/GmodPath.h"

namespace dnv::vista::sdk::tests
{
	struct ExpectedValues
	{
		std::string MaxCode;
		int NodeCount;
	};

	const std::map<VisVersion, ExpectedValues> ExpectedMaxes = {
		{ VisVersion::v3_4a, { "C1053.3112", 6420 } },
		{ VisVersion::v3_5a, { "C1053.3112", 6557 } },
		{ VisVersion::v3_6a, { "C1053.3112", 6557 } },
		{ VisVersion::v3_7a, { "H346.11113", 6672 } },
		{ VisVersion::v3_8a, { "H346.11112", 6335 } } };

	struct TraversalState
	{
		TraversalState( int stopAfter ) : StopAfter( stopAfter ), NodeCount( 0 ) {}
		int StopAfter;
		int NodeCount;
	};

	std::size_t occurrences( const std::vector<const GmodNode*>& parents, const GmodNode& node )
	{
		return static_cast<size_t>(
			std::count_if( parents.begin(), parents.end(),
				[&node]( const GmodNode* p ) {
					return p != nullptr && p->code() == node.code();
				} ) );
	}

	class GmodTests : public ::testing::TestWithParam<VisVersion>
	{
	protected:
		static std::pair<VIS&, const Gmod&> visAndGmod( VisVersion visVersion )
		{
			VIS& vis = VIS::instance();
			const auto& gmod = vis.gmod( visVersion );
			return { vis, gmod };
		}
	};

	//=====================================================================
	// TEST_F
	//=====================================================================

	//----------------------------------------------
	// TEST_F Passing :)
	//----------------------------------------------

	TEST_F( GmodTests, Test_Gmod_Node_Equality )
	{
		auto [vis, gmod] = visAndGmod( VisVersion::v3_4a );

		const auto& node1 = gmod["400a"];
		const auto& node2 = gmod["400a"];

		EXPECT_EQ( node1, node2 );
		EXPECT_EQ( &node1, &node2 );

		auto node3 = node1.withLocation( "1" );
		EXPECT_NE( node1, node3 );
		EXPECT_NE( &node1, &node3 );
	}

	TEST_F( GmodTests, Test_Gmod_Node_Types )
	{
		auto [vis, gmod] = visAndGmod( VisVersion::v3_4a );

		std::unordered_set<std::string> types;
		for ( auto it = gmod.begin(); it != gmod.end(); ++it )
		{
			const auto& node = *it;
			types.insert( node.metadata().category() + " | " + node.metadata().type() );
		}

		EXPECT_FALSE( types.empty() );
	}

	TEST_F( GmodTests, Test_Product_Selection )
	{
		auto [vis, gmod] = visAndGmod( VisVersion::v3_4a );

		const auto& node = gmod["CS1"];
		EXPECT_TRUE( node.isProductSelection() );
	}

	TEST_F( GmodTests, Test_Normal_Assignments )
	{
		auto [vis, gmod] = visAndGmod( VisVersion::v3_4a );

		const auto& node1 = gmod["411.3"];
		(void)node1;
		EXPECT_TRUE( node1.productType() != nullptr );
		EXPECT_TRUE( node1.productSelection() == nullptr );

		const auto& node2 = gmod["H601"];
		(void)node2;
		EXPECT_TRUE( node2.productType() == nullptr );
	}

	TEST_F( GmodTests, Test_Node_With_Product_Selection )
	{
		auto [vis, gmod] = visAndGmod( VisVersion::v3_4a );

		const auto& node1 = gmod["411.2"];
		(void)node1;
		EXPECT_TRUE( node1.productSelection() != nullptr );
		EXPECT_TRUE( node1.productType() == nullptr );

		const auto& node2 = gmod["H601"];
		(void)node2;
		EXPECT_TRUE( node2.productSelection() == nullptr );
	}

	//----------------------------------------------
	// TEST_F Failing :(
	//----------------------------------------------

	TEST_F( GmodTests, Test_Full_Traversal )
	{
		auto [vis, gmod] = visAndGmod( VisVersion::v3_4a );

		int pathCount = 0;
		const int maxExpected = Gmod::TraversalOptions::DEFAULT_MAX_TRAVERSAL_OCCURRENCE;
		size_t maxOccurrence = 0;

		bool completed = gmod.traverse(
			[&]( const std::vector<const GmodNode*>& parents, const GmodNode& node ) {
				EXPECT_TRUE( parents.empty() || ( parents[0] != nullptr && parents[0]->isRoot() ) );

				if ( std::any_of( parents.begin(), parents.end(),
						 []( const GmodNode* p ) { return p != nullptr && p->code() == "HG3"; } ) ||
					 node.code() == "HG3" )
				{
					pathCount++;
				}

				bool skipOccurrenceCheck = Gmod::isProductSelectionAssignment(
					parents.empty() ? nullptr : parents.back(), &node );

				if ( skipOccurrenceCheck )
					return Gmod::TraversalHandlerResult::Continue;

				size_t occ = occurrences( parents, node );
				if ( occ > maxOccurrence )
				{
					maxOccurrence = occ;
				}

				return Gmod::TraversalHandlerResult::Continue;
			} );

		EXPECT_EQ( maxExpected, maxOccurrence );
		EXPECT_TRUE( completed );
		EXPECT_GT( pathCount, 0 );
	}

	TEST_F( GmodTests, Test_Full_Traversal_With_Options )
	{
		auto [vis, gmod] = visAndGmod( VisVersion::v3_4a );

		const int maxExpected = 2;
		size_t maxOccurrence = 0;

		Gmod::TraversalOptions options;
		options.maxTraversalOccurrence = maxExpected;

		bool completed = gmod.traverse(
			[&]( const std::vector<const GmodNode*>& parents, const GmodNode& node ) {
				bool skipOccurrenceCheck = Gmod::isProductSelectionAssignment(
					parents.empty() ? nullptr : parents.back(), &node );

				if ( skipOccurrenceCheck )
				{
					return Gmod::TraversalHandlerResult::Continue;
				}

				size_t occ = occurrences( parents, node );
				if ( occ > maxOccurrence )
				{
					maxOccurrence = occ;
				}

				return Gmod::TraversalHandlerResult::Continue;
			},
			options );

		EXPECT_EQ( maxExpected, maxOccurrence );
		EXPECT_TRUE( completed );
	}

	TEST_F( GmodTests, Test_Partial_Traversal )
	{
		auto [vis, gmod] = visAndGmod( VisVersion::v3_4a );

		TraversalState state( 5 );

		bool completed = gmod.traverse(
			[&state]( const std::vector<const GmodNode*>& parents, [[maybe_unused]] const GmodNode& node ) {
				EXPECT_TRUE( parents.empty() || ( parents[0] != nullptr && parents[0]->isRoot() ) );
				if ( ++state.NodeCount == state.StopAfter )
					return Gmod::TraversalHandlerResult::Stop;
				return Gmod::TraversalHandlerResult::Continue;
			} );

		EXPECT_EQ( state.StopAfter, state.NodeCount );
		EXPECT_FALSE( completed );
	}

	TEST_F( GmodTests, Test_Full_Traversal_From )
	{
		auto [vis, gmod] = visAndGmod( VisVersion::v3_4a );

		TraversalState state( 0 );
		const auto& startNode = gmod["400a"];
		ASSERT_FALSE( startNode.code().empty() ) << "Start node '400a' not found or invalid.";

		bool completed = gmod.traverse<TraversalState>(
			state,
			startNode,
			[]( TraversalState& state, const std::vector<const GmodNode*>& parents, [[maybe_unused]] const GmodNode& node ) -> Gmod::TraversalHandlerResult {
				EXPECT_TRUE( parents.empty() || ( parents[0] != nullptr && parents[0]->code() == "400a" ) );
				++state.NodeCount;
				return Gmod::TraversalHandlerResult::Continue;
			} );

		EXPECT_TRUE( completed );
		EXPECT_GT( state.NodeCount, 0 );
	}

	//=====================================================================
	// TEST_P
	//=====================================================================

	//----------------------------------------------
	// TEST_P Passing :)
	//----------------------------------------------

	TEST_P( GmodTests, Test_Gmod_Loads )
	{
		auto visVersion = GetParam();
		auto [vis, gmod] = visAndGmod( visVersion );

		const GmodNode* tempNodePtr = nullptr;
		ASSERT_TRUE( gmod.tryGetNode( std::string( "400a" ), tempNodePtr ) ) << "Node '400a' not found in GMOD.";
		ASSERT_NE( tempNodePtr, nullptr ) << "tryGetNode succeeded but pointer is null.";
	}

	TEST_P( GmodTests, Test_Gmod_Lookup )
	{
		auto visVersion = GetParam();
		auto [vis, gmod] = visAndGmod( visVersion );

		auto gmodDto = vis.gmodDto( visVersion );

		{
			std::unordered_set<std::string> seen;
			int counter = 0;

			ASSERT_FALSE( gmodDto.items().empty() );

			for ( const auto& item : gmodDto.items() )
			{
				ASSERT_FALSE( item.code().empty() );
				auto insertResult = seen.insert( item.code() );
				EXPECT_TRUE( insertResult.second ) << "Code: " << item.code();

				const GmodNode* foundNodePtr = nullptr;
				ASSERT_TRUE( gmod.tryGetNode( item.code(), foundNodePtr ) );
				ASSERT_NE( foundNodePtr, nullptr );
				EXPECT_EQ( item.code(), foundNodePtr->code() );
				counter++;
			}
		}

		{
			std::unordered_set<std::string> seen;
			int counter = 0;
			for ( auto it = gmod.begin(); it != gmod.end(); ++it )
			{
				const auto& node = *it;
				ASSERT_FALSE( node.code().empty() );
				auto insertResult = seen.insert( node.code() );
				EXPECT_TRUE( insertResult.second ) << "Code: " << node.code();

				const GmodNode* foundNodePtr = nullptr;
				ASSERT_TRUE( gmod.tryGetNode( node.code(), foundNodePtr ) );
				ASSERT_NE( foundNodePtr, nullptr );
				EXPECT_EQ( node.code(), foundNodePtr->code() );
				counter++;
			}
		}

		const GmodNode* notFoundNodePtr = nullptr;
		EXPECT_FALSE( gmod.tryGetNode( std::string( "ABC" ), notFoundNodePtr ) );
		EXPECT_FALSE( gmod.tryGetNode( std::string( "" ), notFoundNodePtr ) );
		EXPECT_FALSE( gmod.tryGetNode( std::string( "SDFASDFSDAFb" ), notFoundNodePtr ) );
		EXPECT_FALSE( gmod.tryGetNode( std::string( "✅" ), notFoundNodePtr ) );
		EXPECT_FALSE( gmod.tryGetNode( std::string( "a✅b" ), notFoundNodePtr ) );
		EXPECT_FALSE( gmod.tryGetNode( std::string( "ac✅bc" ), notFoundNodePtr ) );
		EXPECT_FALSE( gmod.tryGetNode( std::string( "✅bc" ), notFoundNodePtr ) );
		EXPECT_FALSE( gmod.tryGetNode( std::string( "a✅" ), notFoundNodePtr ) );
		EXPECT_FALSE( gmod.tryGetNode( std::string( "ag✅" ), notFoundNodePtr ) );
	}

	//----------------------------------------------
	// TEST_P Failing :(
	//----------------------------------------------

	TEST_P( GmodTests, Test_Gmod_RootNode_Children )
	{
		auto visVersion = GetParam();
		auto [vis, gmod] = visAndGmod( visVersion );

		const auto& node = gmod.rootNode();

		EXPECT_FALSE( node.children().empty() );
	}

	TEST_P( GmodTests, Test_Gmod_Properties )
	{
		auto visVersion = GetParam();
		auto [vis, gmod] = visAndGmod( visVersion );

		auto expectedIt = ExpectedMaxes.find( visVersion );
		ASSERT_NE( expectedIt, ExpectedMaxes.end() );
		std::string expectedMaxCode = expectedIt->second.MaxCode;

		const GmodNode* minLength = nullptr;
		const GmodNode* maxLength = nullptr;
		const GmodNode* expectedMax = nullptr;

		int nodeCount = 0;
		for ( auto nodeIt = gmod.begin(); nodeIt != gmod.end(); ++nodeIt )
		{
			const auto& node = *nodeIt;
			nodeCount++;

			if ( !node.code().empty() )
			{
				if ( minLength == nullptr || node.code().length() < minLength->code().length() )
					minLength = &node;

				if ( maxLength == nullptr || node.code().length() > maxLength->code().length() )
					maxLength = &node;

				if ( node.code() == expectedMaxCode )
					expectedMax = &node;
			}
		}

		SPDLOG_INFO( "Found min length node: {}", minLength ? minLength->code() : "null" );
		SPDLOG_INFO( "Found max length node: {}", maxLength ? maxLength->code() : "null" );

		if ( expectedMax )
		{
			SPDLOG_INFO( "Using expected max node: {} (length {})",
				expectedMax->code(), expectedMax->code().length() );
			ASSERT_EQ( expectedMax->code().length(), 10 );
			maxLength = expectedMax;
		}
		else
		{
			SPDLOG_ERROR( "Expected max node {} not found!", expectedMaxCode );
		}

		ASSERT_NE( minLength, nullptr );
		ASSERT_NE( maxLength, nullptr );
		EXPECT_EQ( minLength->code().length(), 2 );
		EXPECT_EQ( minLength->code(), "VE" );
		EXPECT_EQ( maxLength->code().length(), 10 );
		EXPECT_EQ( maxLength->code(), expectedMaxCode );

		EXPECT_EQ( nodeCount, expectedIt->second.NodeCount );
	}

	struct MappabilityTestCase
	{
		std::string Code;
		bool Mappable;
	};

	class MappabilityTests : public ::testing::TestWithParam<MappabilityTestCase>
	{
	};

	TEST_P( MappabilityTests, Test_Mappability )
	{
		try
		{
			const auto& testCase = GetParam();

			auto& vis = VIS::instance();
			const auto& gmod = vis.gmod( VisVersion::v3_4a );

			const GmodNode* nodePtr = nullptr;
			if ( !gmod.tryGetNode( testCase.Code, nodePtr ) )
			{
				FAIL() << "Node '" << testCase.Code << "' not found.";
				return;
			}
			ASSERT_NE( nodePtr, nullptr );

			EXPECT_EQ( nodePtr->isMappable(), testCase.Mappable )
				<< "Mappability mismatch for node '" << testCase.Code << "'.";
		}
		catch ( const std::exception& ex )
		{
			FAIL() << "Exception during mappability test: " << ex.what();
		}
		catch ( ... )
		{
			FAIL() << "Unknown exception during mappability test";
		}
	}

	INSTANTIATE_TEST_SUITE_P(
		GmodTests,
		MappabilityTests,
		::testing::Values(
			MappabilityTestCase{ "VE", false },
			MappabilityTestCase{ "300a", false },
			MappabilityTestCase{ "300", true },
			MappabilityTestCase{ "411", true },
			MappabilityTestCase{ "410", true },
			MappabilityTestCase{ "651.21s", false },
			MappabilityTestCase{ "924.2", true },
			MappabilityTestCase{ "411.1", false },
			MappabilityTestCase{ "C101", true },
			MappabilityTestCase{ "CS1", false },
			MappabilityTestCase{ "C101.663", true },
			MappabilityTestCase{ "C101.4", true },
			MappabilityTestCase{ "C101.21s", false },
			MappabilityTestCase{ "F201.11", true },
			MappabilityTestCase{ "C101.211", false } ) );

	INSTANTIATE_TEST_SUITE_P(
		GmodTestSuite,
		GmodTests,
		::testing::Values(
			VisVersion::v3_4a,
			VisVersion::v3_5a,
			VisVersion::v3_6a,
			VisVersion::v3_7a,
			VisVersion::v3_8a ) );
}
